### PR TITLE
feat: tag cleanup job

### PR DESCRIPTION
**Pull request by jrasm91**
- Add a job to remove empty tags
- Add a button in the UI to manually run a few different clean up jobs

| Light | Dark | 
| - | - |
| image | image |
| image | image |
| image| image |

### CODE REVIEW

File: server/src/repositories/tag.repository.ts [hunk 1]
```diff
@@ -174,6 +179,34 @@ export class TagRepository implements ITagRepository {
...
+        }
+      }
+
+      if (ids.size > 0) {
+        await manager.delete(TagEntity, { id: In([...ids]) });
+        this.logger.log(`Deleted ${ids.size} empty tags`);
+      }
+    });
+  }
+
```
Comment (jrasm91) [inline server/src/repositories/tag.repository.ts:209]: @mertalev if you know a better way to do this, feel free to replace it.

**Review by danieldietzler — APPROVED** [review]
Code looks slick 🚀

File: server/src/repositories/tag.repository.ts [hunk 2]
```diff
@@ -174,6 +179,34 @@ export class TagRepository implements ITagRepository {
...
+          this.logger.debug(`Found empty tag: ${tag.id} - ${tag.value}`);
+          ids.add(tag.id);
+        }
+      }
+
+      if (ids.size > 0) {
+        await manager.delete(TagEntity, { id: In([...ids]) });
+        this.logger.log(`Deleted ${ids.size} empty tags`);
+      }
+    });
```
Comment (mertalev) [inline server/src/repositories/tag.repository.ts:207]:
```suggestion
    await this.dataSource.query(`
      delete from tags
      where not exists(
        select 1
        from tag_asset
        where tag_asset."tagsId" = tags.id
      )`);
```

Comment (jrasm91) [reply server/src/repositories/tag.repository.ts:207]: If I have a tag Parent/Child I don't want to delete Parent or Child if Child is connected to an asset. Does this handle that case?

Comment (mertalev) [reply server/src/repositories/tag.repository.ts:207]: Yes, assuming that if the child has a row in tag_asset, the parent also has a row.

Comment (jrasm91) [reply server/src/repositories/tag.repository.ts:207]: The parent doesn't necessarily have a row.

Comment (mertalev) [reply server/src/repositories/tag.repository.ts:207]: Hm, why not? It isn't a parent/child relationship if the parent doesn't always include the child.

**Comment by alextran1502** [comment]
I find the button `Create job` confusing. Maybe we can have a Clean up job card with three buttons corresponding to the three jobs we have here?

**Comment by jrasm91** [comment]
Maybe. I think there are plans to add more manual jobs and potentially trigger all jobs to be run through a separate interface like this. I'm open to different UX though. 

The current problem is that we only have two options - force=true or force=false - and not all of our jobs work with just that option.

Some examples:
- all the clean up jobs
- metadata extraction just needs to be run on videos
- any job just needs to run on assets processed after a certain date
- changing between thumbnail formats for one thumbnail type doesn't need to regenerate everything

Maybe we can work towards separating the queue stats from the way we queue jobs and the types of jobs we can queue.

**Comment by alextran1502** [comment]
@jrasm91 sounds good, we can think of a better way of representing these new jobs in a different PR

**Review by alextran1502 — APPROVED** [review]

File: server/src/repositories/tag.repository.ts (hunk 2 above)
Comment (jrasm91) [reply server/src/repositories/tag.repository.ts:207]: I don't know how to exactly write the not exists query with the current query builder, but I can try to take a look at this in the future.

**Pull request merged**